parser-typechecker/src/Unison/Typechecker.hs

```diff
@@ -405,3 +408,13 @@ wellTyped ppe env term = go <$> runResultT (synthesize ppe Context.PatternMatchC
 -- `forall a b . a -> b -> a` to be different types
 -- equals :: Var v => Type v -> Type v -> Bool
 -- equals t1 t2 = isSubtype t1 t2 && isSubtype t2 t1
+
+-- | Checks if the mismatch between two types is due to a missing delay, if so returns a tag for which type is
+-- missing the delay
+isMismatchMissingDelay :: (Var v) => Type v loc -> Type v loc -> Maybe (Either (Type v loc) (Type v loc))
```

Comment:
@pchiusano  is this a reasonable way to do this? I suppose ideally this would be done inside the typechecker itself, but it seems like a pretty simple type to check against 🤔 

It works in simple cases; but unsurprisingly doesn't seem to work with polymorphic types; e.g. this doesn't trip the hint:

```
y : '(e -> Nat)
y =
  _ = 1
  _ = "one two three"
  (_a -> 2)
```

If adding it correctly to the typechecker is difficult it's still probably to get the hint _sometimes_ rather than not at all; but would be nice if it worked in all cases :)

Reply:
cc @dolio

Reply:
So, here's a possible idea for how to make the typechecker facilitate some of this. Up to you if you want to try out implementing it.

So, we can detect `do` in the typechecker because it turns into `Lam body` where `Var.typeOf (ABT.variable body) = Delay`. We can try to detect `'t` in the typechecker by looking for a check against `() -> t`. So, you can try to partition the checking problem into three cases:

1. Checking `λDelay -> ...` against `() -> t`: this looks fine, so just continue as normal
2. Checking `λDelay -> ...` against something else: this might be a bad `do`, so put something in the error scope that indicates this.
3. Checking something else against `() -> t`: this might be a missing `do`, so again put something in the error scope.

Then you can try to parse the improved error information in some situations. Like, if you got a leaf mismatch with an arrow vs. something else, you could look to see if one of these `do` indicators is upstream, and report something about that instead.

I think with this approach, you might be able to identify the example you mentioned above. I'm not 100% sure that it wouldn't cause some other weird corner cases, though.